Moto server

martindurant opened the pull request:
Changes moto for s3 tests from monkeypatched/mocking to server mode. This allows aiobotocore exceptions to raise correctly, needed for tests to pass; the change is required by the upcoming async release of s3fs, but also works for old sync (botocore) version. A release of s3fs without this change would break pandas tests.

Note: since I now need to pass storage_options to the various s3 calls in the tests to make them pass (giving the endpoint of the moto s3 server), I needed to also plumb storage_options through the excel IO, which was previously missing, and update the excel tests.

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
Pepper storage_options

Commit:
Merge branch 'master' into storage_options

Commit:
Add feather test

Commit:
Merge branch 'master' into storage_options

Commit:
Add CSV and parquet options tests; lint

Commit:
deeper lint

Commit:
more tests

Commit:
blank line

Commit:
attempt relint

Commit:
unused import

Commit:
more order

Commit:
plumb stata and test

Commit:
Add note about storage_options in whatsnew

Commit:
Plumb and test markdown

Commit:
optional markdown

Commit:
remove extraneous

Commit:
more extraneous

Commit:
Add fsspec options error and docstrings

Commit:
fix that

Commit:
black

Commit:
fix it again

Commit:
more lint

Commit:
Merge branch 'master' into storage_options

Commit:
Requested changes

Commit:
Make moto server process instead of monkey

Commit:
Merge branch 'master' into storage_options

Commit:
Update versions

Commit:
Merge branch 'storage_options' into moto_server

Commit:
black and start excel

Commit:
Plumb excel

Commit:
Merge branch 'master' into storage_options

Commit:
Merge branch 'storage_options' into moto_server

Commit:
Merge branch 'master' into moto_server

Commit:
fix merge

Commit:
isort

Commit:
option typo

Commit:
remove moto variable

Commit:
Add flask where there is moto

Apparently moto in server mode needs flask
https://github.com/spulec/moto/blob/master/setup.py#L88

Commit:
specific options for s3

Commit:
skip some; fix unrelated HDF arg order

Commit:
rerun generate_pip_deps

Commit:
try simpler

Commit:
try again

Commit:
Check moto not py

Commit:
Merge branch 'master' into moto_server

martindurant commented:
Good to go.
Does something like this need a whatsnew, when it only affects the tests and won't be seen by users?
I could add whatsnew about the excel integration, although it's pretty niche.

Sorry about the long commit list, it contains the original list of commits from #35381 (i.e., wasn't squashed here, only in master).

jreback commented on pandas/io/excel/_odfreader.py:
```diff
@@ -18,11 +18,13 @@ class _ODFReader(_BaseExcelReader):
     ----------
     filepath_or_buffer: string, path to be parsed or
         an open readable stream.
+    storage_options: dict (optional)
```
need a versionadded tag

martindurant replied:
Put into `pandas.io.excel._base._read_excel_doc`, so shows up for `pd.read_excel` (i.e., users are not expected to access these inner modules/classes)

jreback commented on pandas/io/excel/_openpyxl.py:
```diff
@@ -467,17 +467,21 @@ def write_cells(
 class _OpenpyxlReader(_BaseExcelReader):
-    def __init__(self, filepath_or_buffer: FilePathOrBuffer) -> None:
+    def __init__(
+        self, filepath_or_buffer: FilePathOrBuffer, storage_options=None
```
can you type wiht Optional[StorageOptions]

martindurant replied:
OK

jreback commented on pandas/io/excel/_openpyxl.py:
```diff
@@ -467,17 +467,21 @@ def write_cells(
         ----------
         filepath_or_buffer : string, path object or Workbook
             Object to be parsed.
+        storage_options: dict (optional)
```
same on all of these

martindurant replied:
Typed using StorageOptions

jreback commented on pandas/tests/io/conftest.py:
```diff
@@ -72,12 +76,30 @@ def add_tips_files(bucket_name):
         try:
-            s3 = moto.mock_s3()
-            s3.start()
+            import shlex
```
can you add some comments on what you are doing

martindurant replied:
done

jreback replied:
you can move the imports to the top of the file that can (subprocess, time shlex)

import requests can be outside the try

jreback commented on pandas/tests/io/parser/test_network.py:
```diff
@@ -52,6 +52,9 @@ def tips_df(datapath):
     return read_csv(datapath("io", "data", "csv", "tips.csv"))
 
 
+s3so = dict(client_kwargs={"endpoint_url": "http://127.0.0.1:5555/"})
```
can we make these a fixture so don't have to define everywhere

martindurant replied:
done

Commit:
Suggestions

Commit:
maybe mypy fix

martindurant commented:
@jreback : went green again

jreback reviewed (changes requested):
looks good, comment on the impl & can you add this issue number to the whatsnew note where you added storage options; i think maybe need to update as now excel works

Commit:
Move fsspec fixture imports; add whatsnew note

martindurant commented:
I didn't understand "comment on the impl" - but I did move the imports and add a sentence to whatsnew for excel.

jreback commented:
lgtm. @TomAugspurger if any comments.

TomAugspurger commented:
I see one environment where there's moto but not flask, which results in skips like https://travis-ci.org/github/pandas-dev/pandas/jobs/717300240#L4633.

- ci/deps/azure-37-locale.yaml

Also, at least some of the tests are being skipped for other builds.

- https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=40289&view=logs&j=4b4d380c-92b5-58cf-5b80-2c6c9b1d5805&t=e685736b-3705-5348-f648-7de631559230&l=2417 (moto too old)
- azure-38-locale: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=40289&view=logs&j=b1c7b65e-b3ce-541a-7fd5-29b4ba56ce18&t=f832e1c9-f124-52e9-664c-6040860879fb&l=2159 (no s3fs)
- azure-37-locale: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=40289&view=logs&j=a3a13ea8-7cf0-5bdb-71bb-6ac8830ae35c&t=99777837-3d45-5d3a-4778-aac6f4679b04&l=2217 (no s3fs)
- travis-37-cov: https://travis-ci.org/github/pandas-dev/pandas/jobs/717300242#L4113 (moto too old)

I think we should take the files that have s3fs and ensure that they have new enough versions of moto and s3fs. That would be

* ci/deps/azure-37-slow.yaml
* ci/deps/azure-windows-37.yaml
* ci/deps/travis-37-cov.yaml
* ci/deps/travis-37.yaml

TomAugspurger commented on pandas/io/excel/_base.py:
```diff
@@ -837,14 +850,16 @@ class ExcelFile:
 
-    def __init__(self, path_or_buffer, engine=None):
+    def __init__(
+        self, path_or_buffer, storage_options: StorageOptions = None, engine=None
```
Put storage_options after engine, for users passing that as a position?

martindurant replied:
done

TomAugspurger commented on pandas/tests/io/conftest.py:
```diff
@@ -56,6 +66,9 @@ def s3_resource(tips_file, jsonl_file, feather_file):
         os.environ.setdefault("AWS_SECRET_ACCESS_KEY", "foobar_secret")
 
         moto = pytest.importorskip("moto")
```
```suggestion
        moto = pytest.importorskip("moto", minversion="1.3.14")
```

martindurant replied:
I didn't know you could do that, would have saved me keystrokes in a lot of places

martindurant replied:
done

TomAugspurger commented on pandas/tests/io/conftest.py:
```diff
@@ -1,4 +1,8 @@
+from distutils.version import LooseVersion
```
```suggestion
```

martindurant replied:
done

TomAugspurger commented on pandas/tests/io/conftest.py:
```diff
@@ -56,6 +66,9 @@ def s3_resource(tips_file, jsonl_file, feather_file):
 
         moto = pytest.importorskip("moto")
+        pytest.importorskip("flask")  # server mode needs flask too
+        if LooseVersion(moto.__version__) < LooseVersion("1.3.14"):
+            pytest.skip("Moto too old")
```
```suggestion
```

martindurant replied:
done

TomAugspurger commented on pandas/tests/io/parser/test_network.py:
```diff
@@ -56,50 +56,62 @@ def tips_df(datapath):
 class TestS3:
     @td.skip_if_no("s3fs")
-    def test_parse_public_s3_bucket(self, tips_df):
+    def test_parse_public_s3_bucket(self, tips_df, s3so):
```
Why does this test (and similar ones) need to change? I understand the changes to the moto-based ones, but not these which (I think) aren't using moto.

martindurant replied:
`@pytest.mark.usefixtures("s3_resource")` a few lines above was installing the mocks in the previous version. I don't know what the intent was, but these tests never touched real S3 (and there is no public "pandas-test" bucket).

TomAugspurger replied:
Ahh gotcha. Thanks!

TomAugspurger reviewed (commented):
Thanks for working on this. A few questions.

Commit:
responses

Commit:
relint

Commit:
update moto deps

Commit:
latest on windows env

Commit:
Add kwargs

These seem to be extra tests that weren't there before?

Commit:
try in win-py38 env

Commit:
Merge branch 'master' into moto_server

Commit:
Env only

Revert all code changes, to see if environments only are responsibble
for failures

Commit:
Revert - reintroduce code

Commit:
Skip test on win

Commit:
not "reason" in skip

(that would be for skipif)

Commit:
typo

Commit:
Merge branch 'master' into moto_server

Commit:
Fewer moto server processes

Commit:
Dumb

Commit:
With exceptions

TomAugspurger commented:
@martindurant the old pyarrow seems to be coming from defaults. Perhaps only conda-forge has new enough packages for us.


So maybe make sure that all the pyarrows have `>=0.12.0` (or whatever you need). That'll get us what we need as long as `conda-forge` is one of the channels in the env.

martindurant commented:
Interestingly, travis-37-locale doesn't have pyarrow listed at all (but the newest version on defaults is 0.16.0)

Commit:
Skip old pyarrow; remove moto where no s3fs in env

Commit:
sign error

Commit:
Add arrow to env that was getting it anyway (but an old one)

Commit:
Add reasonable timeouts

Commit:
lint

Commit:
small revert

TomAugspurger reviewed (approved):
LGTM, will merge later today.
